head_first_design_patterns/tbruinem/json/src/parser.cpp

```diff
@@ -0,0 +1,196 @@
+#include "parser.hpp"
+#include "parsed_json.hpp"
+#include <stdexcept>
+#include <assert.h>
+
+namespace json
+{
+
+static bool IsWhitespace(char c) {
+	return c == ' ' || c == '\t' || c == '\n';
+}
+
+void Parser::UpdateState(ParseState state)
+{
+	this->previous_state = current_state;
+	current_state = state;
+}
+
+StateResult	Parser::StateOut(char c) {
+	//! Still out
+	if (IsWhitespace(c)) {
+		return StateResult::CONTINUE;
+	}
+	//! Within an array, we only have values
+	if (parent == ElementType::JSON_ARRAY) {
+		product = new ParsedJSON();
+		UpdateState(ParseState::VALUE);
+		if (c == '"') {
+			product->SetType(ElementType::JSON_STRING);
+			return StateResult::CONTINUE;
+		}
+		return StateValue(c);
+	}
+	if (c != '"') {
+		return StateResult::ERROR;
+	}
+	//! Name encountered
+	UpdateState(ParseState::WORD);
+	product = new ParsedJSON();
+	return StateResult::CONTINUE;
+}
+
+StateResult Parser::StateWord(char c)
+{
+	assert(product);
+	//! name finished
+	if (c == '"') {
+		product->SetName(temporary);
+		temporary.clear();
+		UpdateState(ParseState::IN);
+		return StateResult::CONTINUE;
+	}
+	temporary += c;
+	return StateResult::CONTINUE;
+}
+
+static bool	IsInteger(const string& str) {
+	if (str.empty()) {
+		return false;
+	}
+	size_t i = 0;
+	if (str[i] == '-') {
+		i++;
+	}
+	return str.find_first_not_of("0123456789", i) == string::npos;
+}
+
+void Parser::ConvertValue() {
+	if (temporary == "false" || temporary == "true") {
+		product->SetType(ElementType::JSON_BOOL);
+		product->SetValueBoolean(temporary == "true");
+	}
+	else if (temporary == "null") {
+		product->SetType(ElementType::JSON_NULL);
+		product->SetValueNULL();
+	}
+	else if (IsInteger(temporary)) {
+		product->SetType(ElementType::JSON_INTEGER);
+		product->SetValueNumber(stoi(temporary));
+	}
+	else {
+		throw std::runtime_error("Could not convert value '" + temporary + "'");
+	}
+}
+
+StateResult Parser::StateValue(char c)
+{
+	assert(product);
+	//No change yet
+	if (IsWhitespace(c)) {
+		//! Value is finished, no whitespace can be in the middle of a value
+		if (product->GetType() != ElementType::JSON_STRING && !temporary.empty()) {
+			ConvertValue();
+			assert(product && product->GetType() != ElementType::UNKNOWN);
+			return StateResult::DONE;
+		}
+		UpdateState(ParseState::VALUE);
+		return StateResult::CONTINUE;
+	}
+	auto product_type = product->GetType();
+	switch (product_type) {
+		case ElementType::JSON_ARRAY:
+		case ElementType::JSON_OBJECT: {
+			if (c == ',') {
```

Comment:
How would you handle characters that are expected in the future but encountered at the wrong iteration i.e `{,,,,,"name": "value"}` would be considered valid?

Reply:
True.. I need to check if they follow a new child